File: binsifter/core/ai_export.py
"""AI-ready export - formats one file's already-extracted BinSifter findings
into a compact Markdown document and a JSON object, for handing to whatever
AI the analyst wants to use: paste the Markdown into a cloud chat interface
(Claude.ai, ChatGPT, etc.), or feed the JSON to a script hitting a local
model's API.

Added 2026-08-08, right after the AI-assisted-triage prototype
(prototype_ollama_triage.py) was concluded NOT viable on real hardware -
running local inference from inside BinSifter itself caused a full GPU/
display lockup on the test machine (see TODO.md's "AI-assisted triage
exploration" section for the full story). This module is the safer
follow-up: BinSifter NEVER runs or calls out to any AI itself here, cloud
or local - it only FORMATS data that's already been computed. There is
zero inference risk, zero network calls, zero new attack surface, because
this is nothing more than a specialized report writer sitting next to
report.py.

Reuses the same "only include fields that actually carry signal" design as
prototype_ollama_triage.py's _compact_row() - empty/zero/False fields are
dropped rather than padding the output with a wall of "None"/"" noise, and
long fields (raw capa output in particular) are truncated so a single
file's export stays a reasonable size to pack into a chat prompt's context.

Deliberately excludes: Status/Progress/Added (internal scan-state
bookkeeping, not relevant to an AI trying to reason about intent), raw
cluster ID numbers (SsdeepClusterId/ImphashClusterId - meaningless without
cross-referencing other files in the same run, unlike ClusterSize which is
self-contained).
"""

from __future__ import annotations

import json
from dataclasses import asdict
from datetime import datetime, timezone
from pathlib import Path

from binsifter.core.models import FileRecord

_TRUNCATE_AT = 4000  # generous vs. prototype_ollama_triage.py's 800 - this
                      # export is meant to be read by a human or pasted into
                      # a large-context chat model, not packed alongside 5+
                      # other files' findings in a single tight API prompt.

_DISCLAIMER = (
    "This document contains only automated findings BinSifter already "
    "extracted for this file - no AI analysis has been run on it. Any "
    "conclusions an AI draws from the data below are a hypothesis for "
    "further investigation, not a detection."
)


def _truncate(value: str) -> str:
    if len(value) <= _TRUNCATE_AT:
        return value
    return value[:_TRUNCATE_AT] + f"\n...(truncated, {len(value)} chars total)"


def _windows_basename(path: str) -> str:
    """FileRecord.Path values are always Windows paths (backslash
    separators) - both variants only ever run on Windows. Deliberately NOT
    using pathlib.Path(...).name here: on a POSIX host (this project's dev
    sandbox, or anyone's CI), pathlib treats backslashes as plain
    characters rather than separators, so Path("C:\\a\\b.exe").name would
    return the WHOLE string instead of "b.exe". Plain string splitting
    gives the same, correct answer on every OS this ever actually runs on.
    """
    return path.replace("\\", "/").rsplit("/", 1)[-1]


def _windows_stem(path: str) -> str:
    name = _windows_basename(path)
    return name.rsplit(".", 1)[0] if "." in name else name


def compact_record(record: FileRecord) -> dict:
    """Build a dict of only the fields worth an AI's attention - empty
    string, zero, False, and -1 (BinSifter's "not computed" sentinel used
    by Entropy/FlossStringCount/SsdeepClusterId) are all dropped rather
    than included as noise. Path/MD5/SHA1 are always kept even if somehow
    empty, since they're the file's identity, not a finding about it.
    Disposition is always kept even at its "Untriaged" default - unlike the
    other sentinels, that's a real, meaningful state (nobody has assessed
    this file yet), not a "not computed" placeholder.
    """
    always_keep = {"Path", "MD5", "SHA1", "Disposition"}
    # Per-field "not computed" sentinels that aren't blank/zero/False/-1 and
    # so wouldn't be caught by the generic check below - YaraSeverity's
    # "Unknown" default means the same thing YaraSeverityScore's -1 does
    # (no YARA hit worth assessing), it's just spelled as a word instead of
    # a number.
    field_sentinels = {"YaraSeverity": "Unknown"}
    raw = asdict(record)
    # Added is a datetime, not JSON-serializable as-is and not useful to an
    # AI reasoning about file behavior anyway - drop unconditionally.
    raw.pop("Added", None)
    raw.pop("Progress", None)
    raw.pop("Status", None)
    raw.pop("SsdeepClusterId", None)
    raw.pop("ImphashClusterId", None)

    out: dict = {}
    for key, value in raw.items():
        if key in always_keep:
            pass
        elif value in ("", 0, False, -1, None):
            continue
        elif key in field_sentinels and value == field_sentinels[key]:
            continue
        if isinstance(value, str):
            value = _truncate(value)
        out[key] = value
    return out


def build_json(record: FileRecord) -> dict:
    """JSON-serializable dict for one file, with a `_meta` block explaining
    what this is (and isn't) for anyone/anything consuming it without the
    surrounding context this module's docstring provides.
    """
    return {
        "_meta": {
            "generated_by": "BinSifter",
            "generated_at": datetime.now(timezone.utc).isoformat(),
            "note": _DISCLAIMER,
        },
        "findings": compact_record(record),
    }


def build_markdown(record: FileRecord) -> str:
    """One self-contained Markdown document for a single file - readable on
    its own, and meant to be pasted directly into a cloud AI chat interface
    without any additional explanation needed from the analyst.
    """
    data = compact_record(record)
    name = _windows_basename(record.Path) or record.Path
    lines: list[str] = [f"# BinSifter finding: {name}", ""]

    lines.append(f"**Path:** `{data.get('Path', record.Path)}`")
    hash_bits = [f"{label} `{data[key]}`" for label, key in
                 (("MD5", "MD5"), ("SHA1", "SHA1"), ("ssdeep", "SSDEEP")) if data.get(key)]
    if hash_bits:
        lines.append(f"**Hashes:** {' · '.join(hash_bits)}")
    lines.append("")

    def section(title: str, rows: list[tuple[str, object]]) -> None:
        present = [(label, value) for label, value in rows if value is not None]
        if not present:
            return
        lines.append(f"## {title}")
        for label, value in present:
            lines.append(f"- {label}: {value}")
        lines.append("")

    section("Signature", [
        ("Status", data.get("SignatureStatus")),
        ("Signer", data.get("SignerName")),
    ])
    # YaraSeverity ("Unknown" default) and YaraSeverityScore (-1 default,
    # dropped by compact_record as a "not computed" sentinel) can each
    # survive compaction independently of the other, so build this line from
    # whichever half is actually present instead of assuming both travel
    # together.
    severity_value = None
    if "YaraSeverity" in data:
        score_part = f" (score {data['YaraSeverityScore']})" if "YaraSeverityScore" in data else ""
        severity_value = f"{data['YaraSeverity']}{score_part}"

    section("YARA", [
        ("Hits", data.get("YaraHitCount")),
        ("Severity", severity_value),
        ("Rules matched", data.get("YaraMatches")),
        ("ATT&CK techniques", data.get("YaraAttackTechniques")),
    ])
    section("capa", [
        ("Eligible", data.get("CapaEligible")),
        ("Detections", data.get("CapaDetectionCount")),
        ("Possible false negative", data.get("PossibleFalseNegative")),
        ("Shellcode format", data.get("CapaShellcodeFormat")),
    ])
    if data.get("CAPAOutput"):
        lines.append("### Raw capa output")
        lines.append("```")
        lines.append(data["CAPAOutput"])
        lines.append("```")
        lines.append("")
    section("ssdeep / imphash clustering", [
        ("ssdeep cluster size", data.get("SsdeepClusterSize")),
        ("ssdeep high similarity to another file", data.get("SsdeepHasHighSimilarity")),
        ("ssdeep previously seen (prior run)", data.get("SsdeepPreviouslySeen")),
        ("ssdeep matches", data.get("SsdeepMatches")),
        ("Imphash", data.get("Imphash")),
        ("Imphash cluster size", data.get("ImphashClusterSize")),
        ("Rich header hash", data.get("RichHash")),
    ])
    ioc_count = data.get("IocCount")
    if data.get("ExtractedIOCs"):
        lines.append(f"## Extracted IOCs{f' ({ioc_count})' if ioc_count else ''}")
        for ioc in str(data["ExtractedIOCs"]).split("; "):
            lines.append(f"- {ioc}")
        lines.append("")
    section("Other", [
        ("Entropy", data.get("Entropy")),
        ("FLOSS string count", data.get("FlossStringCount")),
        ("Packer detected", data.get("PackerDetected")),
        ("Compiler", data.get("Compiler")),
        ("Reputation status", data.get("ReputationStatus")),
        ("Reputation source", data.get("ReputationSource")),
        ("Disposition", data.get("Disposition")),
        ("Source archive", data.get("SourceArchive")),
        ("Scan error", data.get("Error")),
    ])

    lines.append("---")
    lines.append(f"_Generated by BinSifter on {datetime.now(timezone.utc).strftime('%Y-%m-%d')}. {_DISCLAIMER}_")
    return "\n".join(lines)


def export_file(record: FileRecord, output_dir: Path) -> tuple[Path, Path]:
    """Write both the Markdown and JSON export for one file into
    output_dir, named by SHA1 (falling back to the file's own stem if no
    SHA1 is available - e.g. a file that errored before hashing) - same
    naming convention _launch_ghidra() already uses for project folders,
    so exports and Ghidra projects for the same file are easy to
    cross-reference by name. Returns (markdown_path, json_path).
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = f"BinSifter_{record.SHA1}" if record.SHA1 else f"BinSifter_{_windows_stem(record.Path)}"

    md_path = output_dir / f"{stem}.md"
    md_path.write_text(build_markdown(record), encoding="utf-8")

    json_path = output_dir / f"{stem}.json"
    json_path.write_text(json.dumps(build_json(record), indent=2), encoding="utf-8")

    return md_path, json_path

File: ingot/crates/ingot-core/src/ai_export.rs
//! AI-ready export - port of `binsifter.core.ai_export`.
//!
//! Formats one file's **already-computed** findings into a compact Markdown
//! document and a JSON object for the analyst to hand to whatever AI they
//! choose. Ingot never runs or calls any AI here - this is a specialised
//! report writer, nothing more.
//!
//! Empty / zero / `false` / sentinel fields are dropped rather than padded
//! with noise; `CapaOutput` is truncated so one file's export stays a
//! pasteable size.

use std::path::{Path, PathBuf};

use chrono::Utc;
use serde_json::{json, Value};

use crate::model::FileRecord;

const TRUNCATE_AT: usize = 4000;

const DISCLAIMER: &str = "This document contains only automated findings BinSifter already extracted for this file - no AI analysis has been run on it. Any conclusions an AI draws from the data below are a hypothesis for further investigation, not a detection.";

fn truncate(s: &str) -> String {
    if s.chars().count() <= TRUNCATE_AT {
        return s.to_string();
    }
    let head: String = s.chars().take(TRUNCATE_AT).collect();
    format!("{head}\n...(truncated, {} chars total)", s.chars().count())
}

/// `"C:\a\b.exe"` / `"/a/b.exe"` -> `"b.exe"` - handles both separators
/// regardless of host OS (record paths come from whichever machine ran the
/// scan).
fn basename(path: &str) -> &str {
    path.rsplit(['\\', '/']).next().unwrap_or(path)
}

fn stem(path: &str) -> &str {
    let name = basename(path);
    name.rsplit_once('.').map(|(s, _)| s).unwrap_or(name)
}

/// The findings worth an AI's attention - drops "not computed" sentinels.
pub fn compact_record(r: &FileRecord) -> Value {
    let mut out = serde_json::Map::new();

    // identity - always kept
    out.insert("path".into(), json!(r.path));
    if let Some(v) = &r.md5 {
        out.insert("md5".into(), json!(v));
    }
    if let Some(v) = &r.sha1 {
        out.insert("sha1".into(), json!(v));
    }
    if let Some(v) = &r.sha256 {
        out.insert("sha256".into(), json!(v));
    }
    out.insert("disposition".into(), json!(r.disposition)); // meaningful even at "Untriaged"

    let mut opt_str = |k: &str, v: &Option<String>| {
        if let Some(v) = v.as_deref().filter(|s| !s.is_empty()) {
            out.insert(k.into(), json!(truncate(v)));
        }
    };
    opt_str("ssdeep", &r.ssdeep);
    opt_str("imphash", &r.imphash);
    opt_str("richHash", &r.rich_hash);
    opt_str("yaraMatches", &r.yara_matches);
    opt_str("yaraAttackTechniques", &r.yara_attack_techniques);
    opt_str("capaOutput", &r.capa_output);
    opt_str("capaShellcodeFormat", &r.capa_shellcode_format);
    opt_str("ssdeepMatches", &r.ssdeep_matches);
    opt_str("error", &r.error);

    let mut nonempty = |k: &str, v: &str| {
        if !v.is_empty() {
            out.insert(k.into(), json!(v));
        }
    };
    nonempty("signatureStatus", &r.signature_status);
    nonempty("signerName", &r.signer_name);
    nonempty("extractedIocs", &r.extracted_iocs);
    nonempty("reputationStatus", &r.reputation_status);
    nonempty("reputationSource", &r.reputation_source);
    nonempty("packerDetected", &r.packer_detected);
    nonempty("compiler", &r.compiler);
    nonempty("sourceArchive", &r.source_archive);

    if r.nsrl_match {
        out.insert("nsrlMatch".into(), json!(true));
    }
    if r.yara_hit_count > 0 {
        out.insert("yaraHitCount".into(), json!(r.yara_hit_count));
    }
    if r.yara_severity != "Unknown" {
        out.insert("yaraSeverity".into(), json!(r.yara_severity));
    }
    if r.yara_severity_score >= 0 {
        out.insert("yaraSeverityScore".into(), json!(r.yara_severity_score));
    }
    if r.capa_eligible {
        out.insert("capaEligible".into(), json!(true));
    }
    if r.capa_detection_count > 0 {
        out.insert("capaDetectionCount".into(), json!(r.capa_detection_count));
    }
    if r.possible_false_negative {
        out.insert("possibleFalseNegative".into(), json!(true));
    }
    if r.entropy >= 0.0 {
        out.insert(
            "entropy".into(),
            json!((r.entropy * 1000.0).round() / 1000.0),
        );
    }
    if r.floss_string_count >= 0 {
        out.insert("flossStringCount".into(), json!(r.floss_string_count));
    }
    if r.ssdeep_cluster_size > 0 {
        out.insert("ssdeepClusterSize".into(), json!(r.ssdeep_cluster_size));
    }
    if r.ssdeep_has_high_similarity {
        out.insert("ssdeepHasHighSimilarity".into(), json!(true));
    }
    if r.ssdeep_previously_seen {
        out.insert("ssdeepPreviouslySeen".into(), json!(true));
    }
    if r.imphash_cluster_size > 0 {
        out.insert("imphashClusterSize".into(), json!(r.imphash_cluster_size));
    }
    if r.ioc_count > 0 {
        out.insert("iocCount".into(), json!(r.ioc_count));
    }

    Value::Object(out)
}

pub fn build_json(record: &FileRecord) -> Value {
    json!({
        "_meta": {
            "generatedBy": "BinSifter Ingot",
            "generatedAt": Utc::now().to_rfc3339(),
            "note": DISCLAIMER,
        },
        "findings": compact_record(record),
    })
}

pub fn build_markdown(record: &FileRecord) -> String {
    let d = compact_record(record);
    let get = |k: &str| d.get(k);
    let s = |k: &str| get(k).and_then(Value::as_str);
    let name = {
        let b = basename(&record.path);
        if b.is_empty() {
            record.path.as_str()
        } else {
            b
        }
    };

    let mut out = format!("# BinSifter finding: {name}\n\n");
    out.push_str(&format!(
        "**Path:** `{}`\n",
        s("path").unwrap_or(&record.path)
    ));
    let hashes: Vec<String> = [("MD5", "md5"), ("SHA1", "sha1"), ("ssdeep", "ssdeep")]
        .iter()
        .filter_map(|(label, key)| s(key).map(|v| format!("{label} `{v}`")))
        .collect();
    if !hashes.is_empty() {
        out.push_str(&format!("**Hashes:** {}\n", hashes.join(" · ")));
    }
    out.push('\n');

    fn section(out: &mut String, title: &str, rows: &[(&str, Option<String>)]) {
        let present: Vec<_> = rows.iter().filter(|(_, v)| v.is_some()).collect();
        if present.is_empty() {
            return;
        }
        out.push_str(&format!("## {title}\n"));
        for (label, value) in present {
            out.push_str(&format!("- {label}: {}\n", value.as_ref().unwrap()));
        }
        out.push('\n');
    }

    // Render values the way Python's f-strings do: bare strings without JSON
    // quotes, and `True`/`False` for booleans (matching Winnow's output).
    let disp = |k: &str| {
        get(k).map(|v| match v {
            Value::Bool(true) => "True".to_string(),
            Value::Bool(false) => "False".to_string(),
            Value::String(s) => s.clone(),
            other => other.to_string(),
        })
    };

    section(
        &mut out,
        "Signature",
        &[
            ("Status", disp("signatureStatus")),
            ("Signer", disp("signerName")),
        ],
    );
    let severity = get("yaraSeverity").and_then(Value::as_str).map(|sev| {
        match get("yaraSeverityScore").and_then(Value::as_i64) {
            Some(score) => format!("{sev} (score {score})"),
            None => sev.to_string(),
        }
    });
    section(
        &mut out,
        "YARA",
        &[
            ("Hits", disp("yaraHitCount")),
            ("Severity", severity),
            ("Rules matched", disp("yaraMatches")),
            ("ATT&CK techniques", disp("yaraAttackTechniques")),
        ],
    );
    section(
        &mut out,
        "capa",
        &[
            ("Eligible", disp("capaEligible")),
            ("Detections", disp("capaDetectionCount")),
            ("Possible false negative", disp("possibleFalseNegative")),
            ("Shellcode format", disp("capaShellcodeFormat")),
        ],
    );
    if let Some(capa_out) = s("capaOutput") {
        out.push_str("### Raw capa output\n```\n");
        out.push_str(capa_out);
        out.push_str("\n```\n\n");
    }
    section(
        &mut out,
        "ssdeep / imphash clustering",
        &[
            ("ssdeep cluster size", disp("ssdeepClusterSize")),
            (
                "ssdeep high similarity to another file",
                disp("ssdeepHasHighSimilarity"),
            ),
            ("ssdeep matches", disp("ssdeepMatches")),
            ("Imphash", disp("imphash")),
            ("Imphash cluster size", disp("imphashClusterSize")),
        ],
    );
    if let Some(iocs) = s("extractedIocs") {
        let count = get("iocCount")
            .and_then(Value::as_i64)
            .map(|c| format!(" ({c})"))
            .unwrap_or_default();
        out.push_str(&format!("## Extracted IOCs{count}\n"));
        for ioc in iocs.split("; ") {
            out.push_str(&format!("- {ioc}\n"));
        }
        out.push('\n');
    }
    section(
        &mut out,
        "Other",
        &[
            ("Entropy", disp("entropy")),
            ("FLOSS string count", disp("flossStringCount")),
            ("Reputation status", disp("reputationStatus")),
            ("Disposition", disp("disposition")),
            ("Source archive", disp("sourceArchive")),
            ("Scan error", disp("error")),
        ],
    );

    out.push_str("---\n");
    out.push_str(&format!(
        "_Generated by BinSifter Ingot on {}. {DISCLAIMER}_\n",
        Utc::now().format("%Y-%m-%d")
    ));
    out
}

/// Write the `.md` + `.json` export for one file into `output_dir`, named by
/// SHA-1 (or the file stem if unhashed). Returns `(markdown_path, json_path)`.
pub fn export_file(record: &FileRecord, output_dir: &Path) -> std::io::Result<(PathBuf, PathBuf)> {
    std::fs::create_dir_all(output_dir)?;
    let stem = match record.sha1.as_deref().filter(|s| !s.is_empty()) {
        Some(sha1) => format!("BinSifter_{sha1}"),
        None => format!("BinSifter_{}", stem(&record.path)),
    };
    let md_path = output_dir.join(format!("{stem}.md"));
    let json_path = output_dir.join(format!("{stem}.json"));
    std::fs::write(&md_path, build_markdown(record))?;
    std::fs::write(
        &json_path,
        serde_json::to_string_pretty(&build_json(record)).unwrap(),
    )?;
    Ok((md_path, json_path))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn drops_sentinels_keeps_real_findings() {
        let mut r = FileRecord::new("C:\\evidence\\bad.exe");
        r.sha1 = Some("abcd".into());
        r.md5 = Some("ef01".into());
        r.entropy = 7.55;
        r.yara_hit_count = 2;
        r.yara_severity = "Critical".into();
        r.yara_severity_score = 95;
        r.signature_status = "NotTrusted".into();
        r.ioc_count = 3;
        r.extracted_iocs = "1.2.3.4; evil.example.com".into();

        let c = compact_record(&r);
        // dropped: floss_string_count (-1), reputation ("" ), capa (0/false)
        assert!(c.get("flossStringCount").is_none());
        assert!(c.get("reputationStatus").is_none());
        assert!(c.get("capaDetectionCount").is_none());
        // kept
        assert_eq!(c["yaraHitCount"], 2);
        assert_eq!(c["yaraSeverity"], "Critical");
        assert_eq!(c["disposition"], "Untriaged"); // always kept
        assert_eq!(c["entropy"], 7.55);

        let md = build_markdown(&r);
        assert!(md.contains("# BinSifter finding: bad.exe"));
        assert!(md.contains("Status: NotTrusted"));
        assert!(md.contains("Severity: Critical (score 95)"));
        assert!(md.contains("- 1.2.3.4"));
        assert!(md.contains("- evil.example.com"));
        assert!(md.contains("no AI analysis has been run"));
    }

    #[test]
    fn export_writes_both_files_named_by_sha1() {
        let dir = tempfile::tempdir().unwrap();
        let mut r = FileRecord::new("/x/sample");
        r.sha1 = Some("deadbeef".into());
        let (md, js) = export_file(&r, dir.path()).unwrap();
        assert!(md.ends_with("BinSifter_deadbeef.md"));
        assert!(js.ends_with("BinSifter_deadbeef.json"));
        let parsed: Value = serde_json::from_str(&std::fs::read_to_string(&js).unwrap()).unwrap();
        assert_eq!(parsed["_meta"]["generatedBy"], "BinSifter Ingot");
        assert!(parsed["findings"]["path"] == "/x/sample");
    }
}
